Here is a table schema for a P2P messaging application. The table contains send/receive message data for the application's 
users. 

The structure is as follows:

Table name: user_messaging

date

sender_id (#id of the message sender)

receiver_id (#id of the message receiver)

Question: What fraction of active users sent messages to at least 9 unique people on March 1, 2018? 

In [90]:
#Importing all libraries
import pandas as pd
import numpy as np

#Reading in the data-set
df = pd.read_csv('https://raw.githubusercontent.com/erood/interviewqs.com_code_snippets/master/Datasets/ddi_message_app_data.csv')
df.head()

date  sender_id  receiver_id
0  2018-03-01          5            2
1  2018-03-01          8            6
2  2018-03-01          1            2
3  2018-03-01          4            8
4  2018-03-01          2            7

In [91]:
#Dataframe information 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 999 entries, 0 to 998
Data columns (total 3 columns):
date           999 non-null object
sender_id      999 non-null int64
receiver_id    999 non-null int64
dtypes: int64(2), object(1)
memory usage: 23.5+ KB


Dataframe information reveals that there are 3 columns and 999 non-zero values

In [92]:
#Unique/Distinct values in the date column 
df['date'].unique()

array(['2018-03-01', '2018-03-02', '2018-03-03'], dtype=object)

In [93]:
#No. of values for each date
df['date'].value_counts()

2018-03-01    459
2018-03-03    291
2018-03-02    249
Name: date, dtype: int64

In [94]:
#Sub-setting the dataframe for 1st March 2018 only
df2=df[df['date']=='2018-03-01']

In [95]:
#Subset df2 information matches with the unique count for '1st-March 2018'
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 459 entries, 0 to 458
Data columns (total 3 columns):
date           459 non-null object
sender_id      459 non-null int64
receiver_id    459 non-null int64
dtypes: int64(2), object(1)
memory usage: 14.3+ KB


In [96]:
#Determining the unique values in the subset data-frame and storing them in a new dataframe b
b=pd.DataFrame(df2['sender_id'].value_counts())

In [97]:
b

sender_id
4          63
1          58
2          50
6          49
5          49
8          43
3          42
10         39
9          35
7          31

In [98]:
b.index

Int64Index([4, 1, 2, 6, 5, 8, 3, 10, 9, 7], dtype='int64')

In [99]:
#Counting the total active users for March 1st 2018
total_active=np.count_nonzero(b.index)
total_active

10

In [100]:
#Looping through the subset dataframe df2 and finding the total no. of users messaging atleast 9 users on March 1st 2018

fr=0  # Initiating an empty variable as a counter to count the total users

for i in b.index: # Looping through b.index since it has all the unique active users
    
    df3=df2[df2['sender_id']==i] # Storing a subset of each sender id into df3
    df3_uniq=df3['receiver_id'].unique() #Storing the unique receiver values for each sender id 
    ct=np.count_nonzero(df3_uniq) #Counting the total unique receivers
    
    if(ct>=9):
        fr=fr+1  #Incrementing the counter if the count of total unique receivers is atleast 9

In [101]:
fr

9

In [102]:
pct_active_march1=(fr/total_active)*100
print("The percentage of active users sending messages to at least 9 unique people on March 1, 2018 are :"
      ,pct_active_march1,"%")

The percentage of active users sending messages to at least 9 unique people on March 1, 2018 are : 90.0 %
